contracts/mockContracts/MockDGXDex.sol

```diff
@@ -0,0 +1,187 @@
+/// @dev a dummy contract that simulates Digix contract for purchase/sell with verified signatures. 
+contract MockDGXDEX is Withdrawable {
+    uint  constant internal G_TO_MG = (10**3);
+    uint  constant internal DGX_DECIMALS = (10**9);
+    
+    ERC20 token;
+    address feedSigner;
+    bool public sigCheck = true;
+    bool public blockCheck = true;
+
+    function MockDGXDEX(ERC20 _token, address _feedSigner, address _admin) public {
+        admin = _admin;
+        token = _token;
+        feedSigner = _feedSigner;
+    }
+    
+    function () public payable {}
+    
+    event Purchase(bool success, uint256 purchasedAmount);
+    event Sell(bool success, uint256 amountWei);
+
+    function setSigCheck (bool _sigCheck) public onlyAdmin {
+        sigCheck = _sigCheck;
+    }
+
+    function setBlockCheck (bool _blockCheck) public onlyAdmin {
+        blockCheck = _blockCheck;
+    }
+
+    function purchase(uint256 block_number,
+                      uint256 nonce,
+                      uint256 weiPerDgxMg,
+                      address signer,
+                      bytes signature) payable public returns (bool success, uint256 purchasedAmount) {
+
+        uint256 tokenAmount;
+
+        if( sigCheck) {
+            bool verified;
+            address actual_signer;
+            
+            (verified, actual_signer) =   
+            verify_signed_price(block_number,
+                                nonce,
+                                weiPerDgxMg,
+                                signer,
+                                signature);
+
+            require(verified);
+        }
+        if(blockCheck) {
+            require((block_number + 5) >= block.number);
+        }
+
+        require((block_number + 5) >= block.number);
+
+        tokenAmount = (msg.value / weiPerDgxMg) * ( DGX_DECIMALS / G_TO_MG);
+        require(token.transfer(msg.sender, tokenAmount));
+
+        success = true;
+        purchasedAmount = tokenAmount;
+        
+        Purchase(success, purchasedAmount);
+        
+    }
+
+    function sell(uint256 amount,
+                  uint256 block_number,
+                  uint256  nonce,
+                  uint256 weiPerDgxMg,
+                  address signer,
+                  bytes signature) public returns (bool success) {
+
+        uint256 amountWei;
+
+        if (sigCheck) {
+            bool verified;
+            address actual_signer;
+            
+            (verified, actual_signer) =   
+            verify_signed_price(block_number,
+                                nonce,
+                                weiPerDgxMg,
+                                signer,
+                                signature);
+
+            require(verified);
+        }
+        if (blockCheck) {
+            require((block_number + 5) >= block.number);
+        }
+
+        amountWei = amount * G_TO_MG * weiPerDgxMg / DGX_DECIMALS;
+        require(token.transferFrom(msg.sender, this, amount));
```

Comment:
can't verify this formula. not sure what each number means.

Reply:
Changed the calculation to be in steps (also on purchase function). Hope it's clearer now.

Purchase:
amountMg = (msg.value / weiPerDgxMg)
amountMgDgxWei = amountMg * DGX_DECIMALS;
amountDgxWei = amountMgDgxWei / GRAM

Sell:
amountDgxWei = amount;
amountMgDgxWei = amountDgxWei * GRAM_TO_MILLIGRAM;
amountWei = amountMgDgxWei * weiPerDgxMg / DGX_DECIMALS;